protocols/autonat/src/v2/server/handler/dial_request.rs

```diff
@@ -0,0 +1,349 @@
+            let data_count = match coder
+                .next()
+                .await
+                .map_err(|_e| HandleFail::InternalError(idx))?
+            {
+                Request::Dial(_) => {
+                    return Err(HandleFail::RequestRejected);
+                }
+                Request::Data(dial_data_response) => dial_data_response.get_data_count(),
+            };
+            rem_data = rem_data.saturating_sub(data_count);
+            *data_amount += data_count;
+        }
+    }
+    let (back_channel, rx) = oneshot::channel();
+    let dial_back_cmd = DialBackCommand {
+        addr,
+        nonce,
+        back_channel,
+    };
+    dial_back_cmd_sender
+        .clone()
+        .send(dial_back_cmd)
+        .await
+        .map_err(|_| HandleFail::DialBack {
+            idx,
+            result: Err(DialBackStatus::DialErr),
+        })?;
+
+    let dial_back = rx.await.map_err(|_e| HandleFail::InternalError(idx))?;
+    if let Err(err) = dial_back {
+        return Err(HandleFail::DialBack {
+            idx,
+            result: Err(err),
+        });
+    }
+    Ok(DialResponse {
+        status: ResponseStatus::OK,
+        addr_idx: idx,
+        dial_status: DialStatus::OK,
+    })
+}
+
+async fn handle_request(
+    stream: impl AsyncRead + AsyncWrite + Unpin,
+    observed_multiaddr: Multiaddr,
+    dial_back_cmd_sender: mpsc::Sender<DialBackCommand>,
+    rng: impl RngCore,
+    all_addrs: &mut Vec<Multiaddr>,
+    tested_addrs: &mut Option<Multiaddr>,
+    data_amount: &mut usize,
+) -> io::Result<()> {
+    let mut coder = Coder::new(stream);
+    let response = handle_request_internal(
+        &mut coder,
+        observed_multiaddr,
+        dial_back_cmd_sender,
+        rng,
+        all_addrs,
+        tested_addrs,
+        data_amount,
+    )
+    .await
+    .unwrap_or_else(|e| e.into());
+    coder.send(Response::Dial(response)).await?;
+    coder.close().await?;
+    Ok(())
+}
+
+async fn start_handle_request(
+    stream: impl AsyncRead + AsyncWrite + Unpin,
+    observed_multiaddr: Multiaddr,
+    client: PeerId,
+    dial_back_cmd_sender: mpsc::Sender<DialBackCommand>,
+    mut status_update_sender: mpsc::Sender<StatusUpdate>,
+    rng: impl RngCore,
+) -> io::Result<()> {
+    let mut all_addrs = Vec::new();
+    let mut tested_addrs = None;
+    let mut data_amount = 0;
+    let result = handle_request(
+        stream,
+        observed_multiaddr,
+        dial_back_cmd_sender,
+        rng,
+        &mut all_addrs,
+        &mut tested_addrs,
+        &mut data_amount,
+    )
+    .await;
+    if tested_addrs.is_none() {
+        tracing::warn!("client violated the protocol");
+        return Err(io::Error::from(io::ErrorKind::InvalidData));
+    }
+    let status_update = StatusUpdate {
+        all_addrs,
+        tested_addr: tested_addrs.unwrap(),
+        client,
+        data_amount,
+        result,
+    };
+    let _ = status_update_sender.send(status_update).await;
```

Comment:
I think it would be simpler if we didn't have the `status_update` channel and instead, the `StatusUpdate` would be the return value of this function.

Reply:
yes you are right